psydac/linalg/kernels.py

```diff
@@ -189,3 +189,134 @@ def stencil2coo_3d_F(A:'T', data:'data', rows:'int64[:]', cols:'int64[:]',
                             data[nnz] = value
                             nnz += 1
     return nnz
+
+
+#Implementation of dot products
+
+def dot_product_1d_float64(v1:'float[:]',v2:'float[:]',pads0:'int64',shift0:'int64',shape0:'int64'):
+    '''kernel for computing the inner product (case two real 1d vectors)
+
+    Parameters
+    ----------
+        v1, v2 : 1d float array
+            Data of the vectors from which we are computing the inner product.
+
+        pads0, shift0, shape0 : int
+            pads, shift and shape of the space of the vectors in the 0 direction.
+            
+    Returns
+    -------
+        result : float containing the results'''
+    result = float(0)
+    for i0 in range(pads0*shift0,shape0-pads0*shift0):
+        result += v1[i0]*v2[i0]
+    return result
```

Comment:
@yguclu For all the implemented kernels, we should initialize the `result` with the good type. Thus, @vcarlier decide to separate complex and real cases to initialize `result` as a scalar. But we could unify the complex and real cases by initializing `result` as a numpy array of size (1) with the line :
`result = np.zeros_like(v1, shape=(1,)`

Thus we could reduce the amount of code but it may be slower. What do you think ?

Also, we could also pass the result as an argument of the kernel and thus we should be able to create only one kernel for both case with another pyccel template for this new argument.

Reply:
@tomcaruso I think it is a very good idea to use a Numpy array of size 1 for reducing code duplication. The resulting code may be slower if we are not careful, because Fortran will use an `allocatable` array which is stored in the heap. To avoid this we should use the [`stack_array`](https://github.com/pyccel/pyccel/blob/devel/tutorial/decorators.md#stack-array) decorator:

```python
import numpy as np

from pyccel.decorators import stack_array


@template('T', ['float[:]', 'complex[:]'])
@stack_array('zeros')
def dot_product_1d(
      v1: 'T',
      v2: 'T',
      pads0 : 'int64',
      shift0: 'int64',
      shape0: 'int64'):

    zeros  = np.zeros_like(v1, shape=(1,))
    result = zeros[0]

    for i0 in range(pads0 * shift0, shape0 - pads0 * shift0):
        result += v1[i0] * v2[i0]

    return result
```

Hopefully the compiler will understand that the `zeros` array is useless and it will optimize it out.

Reply:
Actually there is a much simpler solution as suggested by @EmilyBourne:

```python
from pyccel.decorators import template

@template('T', ['float[:]', 'complex[:]'])
def dot_product_1d(
      v1: 'T',
      v2: 'T',
      pads0 : 'int64',
      shift0: 'int64',
      shape0: 'int64'):

    result = v1[0] - v1[0]   # = 0

    for i0 in range(pads0 * shift0, shape0 - pads0 * shift0):
        result += v1[i0] * v2[i0]

    return result
```

Reply:
Ok in fact we could not fuze both function because the conjugate always return a complex data even if the entry data is real. Thus, the file could not be pyccelise, because the data type is not consistent and managing complex even in real case will slow the code